packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart

```diff
@@ -0,0 +1,101 @@
+
+abstract class DashboardBlocEvents {}
+
+abstract class DashboardBlocStates {
+  BehaviorSubject<Result<Map<dashboard.Widget, dashboard.WidgetItems?>>> get widgets;
+}
+
+class DashboardBloc extends InteractiveBloc implements DashboardBlocEvents, DashboardBlocStates {
+  DashboardBloc(this._account) {
+    unawaited(refresh());
+
+    _timer = TimerBloc().registerTimer(const Duration(seconds: 30), refresh);
+  }
+
+  final Account _account;
+  late final NeonTimer _timer;
+
+  @override
+  BehaviorSubject<Result<Map<dashboard.Widget, dashboard.WidgetItems?>>> widgets = BehaviorSubject();
+
+  @override
+  void dispose() {
+    _timer.cancel();
+    unawaited(widgets.close());
+    super.dispose();
+  }
+
+  @override
+  Future<void> refresh() async {
+    widgets.add(widgets.valueOrNull?.asLoading() ?? Result.loading());
+
+    try {
+      final widgets = <String, dashboard.WidgetItems?>{};
+      final v1WidgetIDs = <String>[];
+      final v2WidgetIDs = <String>[];
+
+      final response = await _account.client.dashboard.dashboardApi.getWidgets();
+
+      for (final widget in response.body.ocs.data.values) {
+        if (widget.itemApiVersions.contains(2)) {
+          v2WidgetIDs.add(widget.id);
+        } else if (widget.itemApiVersions.contains(1)) {
+          v1WidgetIDs.add(widget.id);
+        } else {
+          debugPrint('Widget supports none of the API versions: ${widget.id}');
+        }
+      }
+
+      if (v1WidgetIDs.isNotEmpty) {
+        debugPrint('Loading v1 widgets: ${v1WidgetIDs.join(', ')}');
+
+        final response = await _account.client.dashboard.dashboardApi.getWidgetItems(widgets: v1WidgetIDs);
+        for (final entry in response.body.ocs.data.entries) {
+          widgets[entry.key] = dashboard.WidgetItems(
+            (final b) => b
+              ..items.replace(entry.value)
+              ..emptyContentMessage = ''
+              ..halfEmptyContentMessage = '',
+          );
+        }
+      }
+
+      if (v2WidgetIDs.isNotEmpty) {
+        debugPrint('Loading v2 widgets: ${v2WidgetIDs.join(', ')}');
+
+        final response = await _account.client.dashboard.dashboardApi.getWidgetItemsV2(widgets: v2WidgetIDs);
+        widgets.addEntries(response.body.ocs.data.entries);
+      }
+
+      this.widgets.add(
+            Result.success(
+              widgets.map(
+                (final id, final items) => MapEntry(
+                  response.body.ocs.data.values.find(id),
+                  items,
+                ),
+              ),
+            ),
+          );
+    } catch (e, s) {
+      debugPrint(e.toString());
+      debugPrint(s.toString());
+
+      widgets.add(Result.error(e));
+      return;
+    }
+  }
+}
+
+extension WidgetFind on Iterable<dashboard.Widget> {
+  dashboard.Widget? tryFind(final String? id) => firstWhereOrNull((final widget) => widget.id == id);
+  dashboard.Widget find(final String id) => firstWhere((final widget) => widget.id == id);
+}
```

Comment:
Please move this to `utils` add unit tests.
Should be fairly easy to just copy the ones we already have for this.

Reply:
I was thinking of creating a class Findable which every thing that we `find` can implement. Then we don't need to duplicate these extensions and can just have one for all.

Reply:
I agree that a common interface is better but the class in question is provided by the nextcloud package so it wouldn't help here.

Reply:
Ah true